### ISSUE TITLE
FTS5 in standard macOS 10.13 SQLite

### ISSUE THREAD

**Issue by siegesmund** [groue/GRDB.swift#373]
This is a framework question, not a bug. I'm running macOS 10.13 which ships with SQLite 3.19.3, built with FTS5 enabled. GRDB hides the FTS5 structs and methods behind a preprocessor flag that's only set when using GRDBCustom and following the custom build instructions. Given that the current macOS version ships with a FTS5 capable version of SQLite, why not allow FTS5 functionality in the standard GRDB package?

**Comment by groue** [comment]
Hello @siegesmund,

This is an excellent suggestion! Maybe SQLite has been upgraded in the latest iOS and watchOS as well.

Do you feel like setting up a pull request for this?

**Comment by siegesmund** [comment]
Sure, I can do that.

**Comment by groue** [comment]
Thanks! I'm curious to see which technique can be used.

**Comment by groue** [comment]
Oh, @siegesmund, since it's very nice of you to have a look, I should add:

- Please don't hesitate asking question if you have a hard time crawling the maze of projects and build configurations. I'll guide you.
- Please describe your plans before submitting a pull request, so that we are both sure that you don't spend useless time on this topic.

**Comment by siegesmund** [comment]
You've put far more thought into this than I have. Keying on OS version is not enough for a number of reasons, including that some features in the code are only available in later version (>= 3.20). Is the SQLite version number already available anywhere in GRDB? If not, what about calling sqlite_version() sometime early in the init process and hiding/exposing at runtime?

**Comment by groue** [comment]
> You've put far more thought into this than I have, so I'd love to hear your thoughts.

I'm happy to share.

> Keying on OS version is not enough for a number of reasons, including that some features in the code are only available in later version of SQLite (>= 3.20).

Yes. There are three Xcode projects, and three frameworks: GRDB, GRDBCustomSQLite, and GRDBCipher. They all share the same source code. But a few configuration flags differ. Sometimes SQLite features are enabled via SDK check, sometimes they are enabled via checking if GRDB is built with SQLCipher or a customized SQLite build. It also happens that it's a mix. See, for example: https://github.com/groue/GRDB.swift/blob/v3.1.0/GRDB/Core/Database.swift#L402-L432

> Is the SQLite version number already available anywhere in GRDB?

SQLite C Apis are used throughout GRDB. C #defines are not exposed to Swift.

> If not, what about calling sqlite_version() sometime early in the init process and hiding/exposing at runtime?

Apis can't be hidden/exposed at runtime.

**I strongly prefer using compile-time checks rather than run-time checks.**

Today, FTS5 extensions are not available unless GRDB is compiled with the `SQLITE_ENABLE_FTS5` option. See for example https://github.com/groue/GRDB.swift/blob/v3.1.0/GRDB/FTS/FTS5.swift#L1

This `SQLITE_ENABLE_FTS5` flag is set as described in [Custom SQLite builds](https://github.com/groue/GRDB.swift/blob/v3.1.0/Documentation/CustomSQLiteBuilds.md):

> `GRDBCustomSQLite-USER.xcconfig`: this file lets GRDB know about extra compilation flags, and enables extra GRDB APIs.

I guess enabling FTS5, depending on the target operating system would need to change this. I'm not quite sure yet how. I'd be interested to hear about various possible options. I'm sure there are several.

But wait, that's not finished: starting SQLite [SQLite 3.20.0](https://www.sqlite.org/releaselog/3_20_0.html), FTS5 needs an SQLite feature called "pointer-passing interfaces". Since GRDB currently only supports FTS5 for custom SQLite builds (3.24.0), it uses pointer-passing interfaces. But macOS 10.13, which ships with 3.19.3 (as you say) would need to use the old way to access the FTS5 API. GRDB used to use the old way: the relevant commit is https://github.com/groue/GRDB.swift/commit/10af87a7d02c3495feed54a1c53911f6d697a8f2.

To sum up... bringing FTS5 to everyone requires a little involvement. If you jump in, be ready to learn a lot about Xcode, Swift, the Swift package manager, and SQLite 😅

**Comment by groue** [comment]
I really don't want that you spend too much time only to find out that you're stuck and that I gave you an impossible task. That could be very frustrating, so please remember that you're not in charge of anything, and that you can withdraw at any time. But you've put your finger on a tricky little topic, you may happen to enjoy it :-)

**Comment by siegesmund** [comment]
Yes, I think I understand the basics of the problem. Indeed tricky. :)

> Apis can't be hidden/exposed at runtime.

I didn't mean to suggest hiding/exposing an API at runtime. My comment was a bit imprecise. I'd found the old way of doing it (not by commit, but by downloading the last version of GRDB I found before SQLite 3.20) and was experimenting with using something like ```String(cString: sqlite3_libversion())``` to fetch the version and then use either the pointer passing interface or the older method. Wondering how far a combination of version checking and @available/#available attributes to check against OS might take us? ```String(cString: sqlite3_libversion())``` seems to indeed report the version accurately so far. Do you see a big gotcha with that approach?

**Comment by groue** [comment]
I better understand now. And I'm happy you already did this research, and that we talk about the same thing :-)

If we had to sum up: compiler checks expose available APIs to library user, and behind them runtime checks can perform fine-grain adaptations (examples [1](https://github.com/groue/GRDB.swift/blob/v3.1.0/GRDB/Core/StatementAuthorizer.swift#L138), [2](https://github.com/groue/GRDB.swift/blob/v3.1.0/GRDB/Core/Database%2BSchema.swift#L306)).

> I'm wondering how far a combination of version checking and @available/#available attributes to check against OS might take us?

Towards a solution, I guess.

> Do you see a big gotcha with that approach?

The real difficulty is that we need availability checks for the GRDB framework (the one that uses the SQLite library that ships with the operating system, which can be built from both Xcode and SPM), and checks for customized flags for the GRDBCustomSQLite framework (the one that uses a [custom build](https://github.com/groue/GRDB.swift/blob/v3.1.0/Documentation/CustomSQLiteBuilds.md), and currently requires Xcode). This can reveal difficult to do without code duplication, because the language does not support `#if FLAG || available(...)` compiler statements. I really don't know what's the best way. I don't know either which Swift guru we could ask advice to.

**Comment by siegesmund** [comment]
To pare the problem down a little, would you accept a pull request that checks the sqlite version and uses the pointer passing interface if the version is >= 3.20? This won't solve the bigger problem, but it will make it easier to, for example, use GRDB as a submodule and manually set the SQLITE_ENABLE_FTS5 flag in "other swift flags." It will then only be a matter of setting the flag, and no code modification will be necessary.

**Comment by groue** [comment]
I'd greatly appreciate if both code paths could be automatically tested (see the [Makefile](https://github.com/groue/GRDB.swift/blob/master/Makefile) and [.travis.yml](https://github.com/groue/GRDB.swift/blob/master/.travis.yml)).

It happens that Travis ships [Xcode 9.3 on maxOS 10.13](https://docs.travis-ci.com/user/reference/osx/#OS-X-Version). Unless I'm mistaken, this is exactly the setup you need.

**Comment by groue** [comment]
I'd also greatly appreciate if documentation were updated so that it makes it clear:

- that FTS5 can be made available for the regular GRDB framework
- under which conditions
- with a detailed how-to.

Doing this documentation job would actually help a lot setting up the tests, since we would just have to test that documentation does not lie.

I admit pull requests are not easy here. But when documentation is kept up-to-date and accurate, everybody can know about added features (not only the PR contributors). When documentation is tested, it's easier to maintain the code in the future and to provide support. This library wouldn't have survived since 2015 and Swift 2.2 without its doc and its tests.

**Comment by groue** [comment]
Hello @siegesmund, a week has passed.

If you have a working solution on your side but don't have the time for setting up a pull request that would expose FTS5 to all GRDB users, that's totally fine. But I don't want an issue to remain open when nobody is working on it: will you please tell me if I should have any expectations or not?

**Comment by siegesmund** [comment]
Hi @groue,
I have a working copy running locally; I would like to make this pull request, but time is tight for me at the moment. I will go ahead and close this but will still aim to make the pull request when I have a little more bandwidth. Thanks for the guidance and all the time you invest in GRDB.

**Comment by groue** [comment]
You're welcome, @siegesmund. Meanwhile, I've just asked a question on the Swift forums, to get some advice: https://forums.swift.org/t/c-interoperability-combinations-of-library-and-os-versions/14029

I hope we'll get some interesting inputs!

**Comment by groue** [comment]
It looks like there is no solution that avoids code duplication, unfortunately.

Meanwhile, I assure you that the topic of this issue is not [forgotten](https://github.com/groue/GRDB.swift/blob/development/TODO.md). It's ready to be picked whenever someone has the time to do it.

**Comment by groue** [comment]
@siegesmund, new options for FTS5 are coming. Check the commit above.

**Comment by groue** [comment]
Shipped in 3.3.0-beta1: [How to enable FTS5](https://github.com/groue/GRDB.swift/blob/master/README.md#enabling-fts5-support)

**Comment by dinhvh** [comment]
Any way to enable FTS5 with Swift Package Manager? via Xcode UI?
